orchestrator/modules/actuators/custom_experiments.py

```diff
@@ -11,60 +14,528 @@
+    """This function derives the parameterization of properties and function signature.
+
+    The parameterization of a property is the default value of the corresponding parameter in func_signature
+
+    Parameters:
+    - properties: The properties to derive the parameterization for
+    - func_signature: The function signature to derive the parameterization from
+    Returns:
+    - A dictionary of property identifiers and their default values
+    Exceptions:
+    - ValueError: If a parameterization cannot be"""
+    param_map = {p.name: p for p in func_signature.parameters.values()}
+    results = {}
+    missing = []
+    for prop in properties:
+        param = param_map.get(prop.identifier, None)
+        if param and param.default is not inspect.Parameter.empty:
+            results[prop.identifier] = param.default
+        else:
+            missing.append(prop.identifier)
+    if missing:
+        raise ValueError(f"Parameterization missing for: {missing}")
+    return results
+
+
+def derive_optional_properties_and_parameterization(
+    func: typing.Callable, required_properties: list[ConstitutiveProperty]
+) -> tuple[list[ConstitutiveProperty], dict[str, typing.Any]]:
+    """This function derives the optional properties and their parameterization from the function signature.
+
+    The optional properties are the keyword parameters of the function that are not in required_properties.
+    The parameterization of an optional property is the default value of the corresponding parameter in func_signature.
+
+    Parameters:
+    - func: The function to derive the optional properties and parameterization from
+    - required_properties: The properties that are required input values.
+    Returns:
+    - A tuple. The first element is a list of optional properties, the second element is a dictionary of property identifiers and their default values
+    Exceptions:
+    - ValueError: If a parameterization cannot be derived for any optional property (unexpected)
+    - ValueError: If a domain cannot be inferred for any optional property"""
+    optional_properties = []
+    for param in inspect.signature(func).parameters.values():
+        if param.name in {prop.identifier for prop in required_properties}:
+            continue
+        if (
+            param.kind
+            in (inspect.Parameter.KEYWORD_ONLY, inspect.Parameter.POSITIONAL_OR_KEYWORD)
+            and param.default is not inspect.Parameter.empty
+        ):
+            inferred_prop = _infer_domain_and_property(
+                param.name, param.annotation, param.default
+            )
+            optional_properties.append(inferred_prop)
+
+    return optional_properties, get_parameterization(
+        optional_properties, inspect.signature(func)
+    )
+
+
+def check_parameters_and_infer(
+    _optional_properties: list[ConstitutiveProperty] | None,
+    _parameterization: dict | None,
+    _required_properties: list[ConstitutiveProperty] | None,
+    func,
```

Comment:
It looks a bit weird that the only parameter that can't be none is the last one. Also required properties being after optional ones looks a bit weird (applies also for return values)